watcher/watcher.go

```diff
@@ -0,0 +1,296 @@
+	imagesIDsMap           map[string]bool
+	wlidsMap               map[string]map[string]string // <wlid> : <containerName> : imageID
+	imageIDsMapMutex       *sync.Mutex
+	wlidsMapMutex          *sync.Mutex
+	currentResourceVersion string // current PodList version, used by watcher (https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)
+}
+
+func NewWatchHandler() *WatchHandler {
+	return &WatchHandler{
+		k8sAPI:           k8sinterface.NewKubernetesApi(),
+		imagesIDsMap:     make(map[string]bool),
+		wlidsMap:         make(map[string]map[string]string),
+		imageIDsMapMutex: &sync.Mutex{},
+		wlidsMapMutex:    &sync.Mutex{},
+	}
+}
+
+func (wh *WatchHandler) addToImageIDsMap(imageID string) {
+	wh.imageIDsMapMutex.Lock()
+	defer wh.imageIDsMapMutex.Unlock()
+	if _, ok := wh.imagesIDsMap[imageID]; !ok {
+		wh.imagesIDsMap[imageID] = true
+	}
+}
+
+func (wh *WatchHandler) addTowlidsMap(wlid string, containerName string, imageID string) {
+	wh.wlidsMapMutex.Lock()
+	defer wh.wlidsMapMutex.Unlock()
+	if _, ok := wh.wlidsMap[wlid]; !ok {
+		wh.wlidsMap[wlid] = make(map[string]string)
+		wh.wlidsMap[wlid][containerName] = imageID
+	} else {
+		wh.wlidsMap[wlid][containerName] = imageID
+	}
+}
+
+func (wh *WatchHandler) GetWlidsMap() map[string]map[string]string {
+	return wh.wlidsMap
+}
+
+func (wh *WatchHandler) GetImageIDsMap() map[string]bool {
+	return wh.imagesIDsMap
+}
+
+func (wh *WatchHandler) buildImageIDsMap(pods core1.PodList) {
+	for _, pod := range pods.Items {
+		for _, imageID := range extractImageIDsFromPod(&pod) {
+			if _, ok := wh.imagesIDsMap[imageID]; !ok {
+				wh.addToImageIDsMap(imageID)
+			}
+		}
+	}
+}
+
+func (wh *WatchHandler) buildwlidsMap(pods core1.PodList) {
+	for _, pod := range pods.Items {
+		parentWlid, err := wh.getParentIDForPod(&pod)
+		if err != nil {
+			glog.Warningf("failed to get parent for pod %s: %s", pod.Name, err.Error())
+			continue
+		}
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			wh.addTowlidsMap(parentWlid, containerStatus.Name, GetImageID(containerStatus.ImageID))
+		}
+	}
+}
+
+func (wh *WatchHandler) getParentIDForPod(pod *core1.Pod) (string, error) {
+	pod.TypeMeta.Kind = "Pod"
+	podMarshalled, err := json.Marshal(pod)
+	if err != nil {
+		return "", err
+	}
+	wl, err := workloadinterface.NewWorkload(podMarshalled)
+	if err != nil {
+		return "", err
+	}
+	kind, name, err := wh.k8sAPI.CalculateWorkloadParentRecursive(wl)
+	if err != nil {
+		return "", err
+	}
+	return pkgwlid.GetWLID(utils.ClusterConfig.ClusterName, wl.GetNamespace(), kind, name), nil
+
+}
+
+// list all pods, build imageIDsMap and wlidsMap
+// set current resource version for pod watcher
+func (wh *WatchHandler) Initialize(scanNewWorkloads bool) error {
+	// list all Pods and their image IDs
+	podsList, err := wh.k8sAPI.ListPods("", map[string]string{})
+	if err != nil {
+		return err
+	}
+
+	wh.buildImageIDsMap(*podsList)
```

Comment:
since we are already receiving a pointer from `ListPods` let's pass a pointer type to the buildImageIDsMap and buildwlidsMap